File: src/MAP_Unity/SerializerLib/SerializerLib/Serialization/Types/MemoryTag.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace SerializerLib
{
	namespace MAP
	{
		//Internal serializer class, don't use outside
		using NetworkObject = List<INetworkType>;
		
		public class MemoryTag : INetworkType
		{
			public MemoryTag()
			{
				this.m_instance_name = new List<byte>(Encoding.UTF8.GetBytes("MEMORY-TAG-SYSTEM"));
			}

			public MemoryTag(string name)
			{
				this.m_instance_name = new List<byte>(Encoding.UTF8.GetBytes(name));
			}

			public override List<byte> RawSerialization()
            {
				List<byte> memoryVector = new List<byte>();
				//NOTE: MEMORY TAG DOES NOT HAVE TO PASS HIS TYPE ID (ASSUMPTION DUE TO COMPOSITION)
				memoryVector.Add((byte)(m_instance_name.Count)); //CHANGE THIS CAST of size_t by byte
				foreach (var stringIterator in m_instance_name)
				{
					memoryVector.Add((stringIterator));
				}
				return memoryVector;
			}
			
			public override List<byte> Serialize()
			{
				return RawSerialization();//just the same due to memory tag desing...
			}
			
			
			public override NetworkObject RawDeserialization(byte[] argsMemory)
			{
				byte tagLength = argsMemory[MememoryOffset.OFFSET_1]; //length first pos
				m_instance_name = new List<byte>();
				for (byte i = 0; i < tagLength; i++)
				{
					m_instance_name.Add(argsMemory[MememoryOffset.OFFSET_2 + i]);
				}

				List<INetworkType> objectStructure = new List<INetworkType>
				{
					new MemoryTag(Encoding.ASCII.GetString(m_instance_name.ToArray()))
				};
				return objectStructure;
			}

			public override List<INetworkType> Deserialize(byte[] argsMemory)
			{
				return RawDeserialization(argsMemory);
			}

			public override NetworkType GetNetworkType()
			{
				return NetworkType.TAG;
			}

			public override string GetName()
			{
				return Encoding.ASCII.GetString(m_instance_name.ToArray());
			}

            public override int GetRawSize()
            {
                return m_instance_name.Count + 1;//1 byte extra for the size byte
            }

            public override int GetSize()
            {
                return GetRawSize();
            }

			private List<byte> m_instance_name;
		}
	}
}

File: src/Serializer/src/Types/MemoryTag.cpp
#include "MemoryTag.hpp"

namespace MAP
{
    MemoryTag::MemoryTag() : m_instance_name(std::string("MEMORY-TAG-SYSTEM"))
    {
    }

    MemoryTag::MemoryTag(const std::string &name) : m_instance_name(name)
    {
    }

    MemoryTag::~MemoryTag()
    {
    }

    std::vector<uint8_t> MemoryTag::RawSerialization()
    {
         std::vector<uint8_t> memoryVector(0);
        //NOTE: MEMORY TAG DOES NOT HAVE TO PASS HIS TYPE ID (ASSUMPTION DUE TO COMPOSITION)
        memoryVector.push_back(static_cast<uint8_t>(m_instance_name.length())); //CHANGE THIS CAST of size_t by uint8_t
        for (auto stringIterator : m_instance_name)
        {
            memoryVector.push_back(stringIterator);
        }
        return memoryVector;
    }

    std::vector<uint8_t> MemoryTag::Serialize()
    {
        return RawSerialization();
    }

    std::vector<std::shared_ptr<INetworkType>> MemoryTag::RawDeserialization(std::vector<uint8_t> argsMemory)
    {
        std::vector<std::shared_ptr<INetworkType>> objectStructure;
        uint8_t tagLength = argsMemory[MEM_OFFSET_1]; //lenght first pos
        m_instance_name = "";
        for (uint8_t i = 0; i < tagLength; i++)
        {
            m_instance_name.push_back(argsMemory[MEM_OFFSET_2 + i]);
        }
        objectStructure.push_back(std::make_shared<MAP::MemoryTag>(m_instance_name));
        return objectStructure;
    }

    std::vector<std::shared_ptr<INetworkType>> MemoryTag::Deserialize(const uint8_t *argsMemory)
    {
        std::vector<std::shared_ptr<INetworkType>> objectStructure;
        uint8_t tagLength = argsMemory[MEM_OFFSET_1]; //lenght first pos
        m_instance_name = "";
        for (uint8_t i = 0; i < tagLength; i++)
        {
            m_instance_name.push_back(argsMemory[MEM_OFFSET_2 + i]);
        }
        objectStructure.push_back(std::make_shared<MAP::MemoryTag>(m_instance_name));
        return objectStructure;
    }

    NetworkType MemoryTag::GetType()
    {
        return NetworkType::TAG;
    }

    const char *MemoryTag::GetName()
    {
        return m_instance_name.c_str();
    }

    uint32_t MemoryTag::GetRawSize(){
        return m_instance_name.size() + 1; //1 byte extra for the size byte
    }

    uint32_t MemoryTag::GetSize()
    {
        return GetRawSize();
    }
}